### PR TITLE
test: ssh-debug integration test

### PR DESCRIPTION
**Pull request by yanksyoon**
Applicable spec: None.

### Overview

This is a follow up PR for #176.

### Rationale

In order to guarantee that the components [tmate-ssh-server charm](https://github.com/canonical/tmate-ssh-server-operator), the current github-runner-operator charm and [action-tmate](https://github.com/canonical/action-tmate) work together.

### Juju Events Changes

None.

### Module Changes

None.

### Library Changes

None.

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [ ] The documentation for charmhub is updated.
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

Documentation will be in a separate PR to reduce scope.

### CODE REVIEW

File: src/charm_state.py [hunk 1]
```diff
@@ -309,6 +309,9 @@ def from_charm(cls, charm: CharmBase) -> "State":
         # Convert pydantic object to python object serializable by json module.
         state_dict["proxy_config"] = json.loads(state_dict["proxy_config"].json())
         state_dict["charm_config"] = json.loads(state_dict["charm_config"].json())
+        state_dict["ssh_debug_info"] = (
+            json.loads(state_dict["ssh_debug_info"].json()) if ssh_debug_info else None
```
Comment (arturo-seijas) [inline src/charm_state.py:313]: Did you mean `if ssh_debug_info in state_dict else None`?

Comment (yanksyoon) [reply src/charm_state.py:313]: `ssh_debug_info` is None if it's not available, so either value can be used here :D

File: tests/integration/conftest.py [hunk 2]
```diff
@@ -357,6 +411,31 @@ async def app_juju_storage(
...
+    branch_ref = github_repository.create_git_ref(
+        ref=f"refs/heads/{test_branch}", sha=Repo().head.commit.hexsha
+    )
+
+    def get_branch():
+        """Get newly created branch."""
+        try:
+            branch = github_repository.get_branch(test_branch)
+        except GithubException as err:
+            if err.status == 404:
```
Comment (arturo-seijas) [inline tests/integration/conftest.py:427]: Why this special treatment if not found?

Comment (yanksyoon) [reply tests/integration/conftest.py:427]: There's a lag between branch creation and getting the branch from GitHub, so essentially we're trying a few more times :)

**Review by cbartz — COMMENTED** [review]
Looks good, few comments.

**Review by gtrkiller — DISMISSED** [review]

File: tests/integration/helpers.py [hunk 3]
```diff
@@ -351,5 +355,46 @@ async def deploy_github_runner_charm(
...
+        if is_awaitable:
+            if result := await func():
+                return result
+        else:
+            if result := func():
+                return result
+        time.sleep(check_interval)
+
+    # final check before raising TimeoutError.
+    if is_awaitable:
```
[the remark this answers is not in the document]
Comment (yanksyoon) [reply tests/integration/helpers.py:394]: Python doesn't quite have a do-while loop, the function above is of a do-while style, this just does a final check before raising an error just in case :D
Would I have understood you correctly here?

[the remark this answers is not in the document]
Comment (arturo-seijas) [reply tests/integration/helpers.py:394]: You did, yes. Just ignore my comment

File: tests/integration/test_debug_ssh.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+                    name
+                    for name in zip_ref.namelist()
+                    if "workflow-dispatch-tests/3_Setup tmate session.txt" == name
+                ]
+            )
+        )
+        logs = str(zip_ref.read(tmate_log_filename), encoding="utf-8")
+
+    # ensure ssh connection info printed in logs.
+    logger.info("Logs: %s", logs)
```
[the remark this answers is not in the document]
Comment (yanksyoon) [reply tests/integration/test_debug_ssh.py]: I've refactored `_assert_workflow_run_conclusion` to be used in a more generic way and refactored some internal functions in `charm_metrics_helpers` to be able to be used cross-modules, for example reading logs. Thank you for the suggestion!

**Review by jdkandersson — DISMISSED** [review]

File: tests/integration/test_debug_ssh.py [hunk 5]
```diff
@@ -3,26 +3,28 @@
...
-from github.Workflow import Workflow
 from github.WorkflowRun import WorkflowRun
+from juju.application import Application
 
-from tests.integration.helpers import wait_for
+from tests.integration.charm_metrics_helpers import (
+    _get_job_logs,
+    dispatch_workflow,
+    get_workflow_runs,
+)
```
Comment (cbartz) [inline tests/integration/test_debug_ssh.py:17]: Could you move the definition of these 3 imported functions to `helpers.py`? `charm_metrics_helpers.py` should only contain helpers related to the `charm_metrics` integration tests. Thank you very much!

**Review by cbartz — COMMENTED** [review]
Looks nice, thanks for the refactoring! One comment.

File: tests/integration/test_debug_ssh.py (hunk 5 above)
Comment (yanksyoon) [reply tests/integration/test_debug_ssh.py]: I've moved functions around to prevent circular imports, the dependencies are now in order of imports: helpers > charm_metrics_helpers > test module.

Thank you for the review!

File: tests/integration/helpers.py [hunk 6]
```diff
@@ -361,6 +371,132 @@ async def deploy_github_runner_charm(
     return application
 
 
+def _get_job_logs(job: WorkflowJob) -> str:
```
Comment (cbartz) [inline tests/integration/helpers.py:374]:
```suggestion
def get_job_logs(job: WorkflowJob) -> str:
```

**Review by cbartz — DISMISSED** [review]
Thanks for the refactoring! I have one nitpicky comment.

File: tests/integration/helpers.py (hunk 6 above)
Comment (yanksyoon) [reply tests/integration/helpers.py]: good catch!

**Review by cbartz — APPROVED** [review]

**Review by arturo-seijas — APPROVED** [review]
